File: 3-hard/02302-count-subarrays-with-score-less-than-k/solution.cpp
class Solution {
public:
    long long countSubarrays(vector<int>& nums, long long k) {
        long long count = 0;
        long long total_sum = 0;
        int left = 0;
        int right = 0;
        int nums_len = nums.size();

        while(true) {
            while (left < nums_len && nums[left] >= k) {
                left += 1;
            }

            if (left >= nums_len) {
                break;
            }

            right = left;
            total_sum = 0;

            while (left <= right) {
                while (right <  nums_len && ((total_sum + nums[right]) * (right - left + 1)) < k) {
                    total_sum += nums[right];
                    right += 1;
                }

                if (right == nums_len || nums[right] >= k) {
                    count += ((long long)(right - left) * (right - left + 1)) / 2;
                    left = right;
                    break;
                }

                count += (right - left);
                total_sum -= nums[left];
                left += 1;
            }
        }

        return count;
    }
};

File: 3-hard/02302-count-subarrays-with-score-less-than-k/solution.py
class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        count, total_sum, left, right = (0, 0, 0, 0)
        while True:
            while left != len(nums) and nums[left] >= k:
                left += 1
            if left == len(nums):
                break
            right = left
            total_sum = nums[left]
            while right >= left:
                while right <  len(nums) - 1 and ((total_sum + nums[right + 1]) * (right - left + 2)) < k:
                    right += 1
                    total_sum += nums[right]
                if right == len(nums) - 1 or nums[right+1] >= k:
                    count += ((right - left + 1) * (right - left + 2))//2
                    left = right + 1
                    break
                count += right - left + 1
                total_sum -= nums[left]
                left += 1
        return count
